File: java/src/main/java/org/demo/interceptor/ClientJwtInterceptor.java
package org.demo.interceptor;

import java.nio.charset.StandardCharsets;
import java.security.Key;

import io.grpc.CallOptions;
import io.grpc.Channel;
import io.grpc.ClientCall;
import io.grpc.ClientInterceptor;
import io.grpc.ForwardingClientCall;
import io.grpc.Metadata;
import io.grpc.MethodDescriptor;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;

/**
 * Intercept call before sending to server, append JWT and use compression
 */
public class ClientJwtInterceptor implements ClientInterceptor {

    private static final Metadata.Key<String> AUTHORIZATION_HEADER
        = Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER);

    @Override
    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> methodDescriptor, CallOptions callOptions, Channel channel) {
        return new ForwardingClientCall.SimpleForwardingClientCall<>(
            //add grpc-encoding=gzip header
            channel.newCall(methodDescriptor, callOptions.withCompression("gzip"))) {
            @Override
            public void start(Listener<RespT> responseListener, Metadata headers) {
                if (!headers.containsKey(AUTHORIZATION_HEADER)) {
                    headers.put(AUTHORIZATION_HEADER, getJwtValue());
                }
                super.start(responseListener, headers);
            }
        };
    }

    private String getJwtValue() {
        Key secretKey = Keys.hmacShaKeyFor(System.getenv("JWT_SECRET").getBytes(StandardCharsets.UTF_8));
        return "Bearer " + Jwts.builder()
            .subject("java-client")
            .signWith(secretKey)
            .compact();
    }

}

File: kotlin/src/main/kotlin/org/demo/interceptor/ClientJwtInterceptor.kt
package org.demo.interceptor

import io.grpc.CallOptions
import io.grpc.Channel
import io.grpc.ClientCall
import io.grpc.ClientInterceptor
import io.grpc.ForwardingClientCall
import io.grpc.MethodDescriptor
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import java.nio.charset.StandardCharsets

class ClientJwtInterceptor : ClientInterceptor {

    val authHeader: io.grpc.Metadata.Key<String> =
        io.grpc.Metadata.Key.of("Authorization", io.grpc.Metadata.ASCII_STRING_MARSHALLER)

    override fun <ReqT : Any?, RespT : Any?> interceptCall(
        methodDescriptor: MethodDescriptor<ReqT, RespT>?,
        callOptions: CallOptions?,
        channel: Channel?
    ): ClientCall<ReqT, RespT> {
        return object : ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
            //add grpc-encoding=gzip header
            channel?.newCall(methodDescriptor, callOptions?.withCompression("gzip"))
        ) {
            override fun start(responseListener: Listener<RespT>, headers: io.grpc.Metadata) {
                if (!headers.containsKey(authHeader)) {
                    headers.put(authHeader, getJwtValue())
                }
                super.start(responseListener, headers)
            }
        }
    }

    fun getJwtValue(): String {
        val secretKey = Keys.hmacShaKeyFor(System.getenv("JWT_SECRET").toByteArray(StandardCharsets.UTF_8))
        return "Bearer " + Jwts.builder()
            .subject("kotlin-client")
            .signWith(secretKey)
            .compact()
    }

}
